packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts

```diff
@@ -0,0 +1,308 @@
+        ? amountToSatoshi(token.balance!, token.decimals)
+        : undefined;
+    if (output.type === 'send-max' || output.type === 'send-max-noaddress') {
+        max = availableTokenBalance || calculateMax(availableBalance, feeInLamports);
+        amount = max;
+    } else {
+        amount = output.amount;
+    }
+
+    // total SOL spent (amount + fee), in case of SPL token only the fee
+    const totalSpent = new BigNumber(calculateTotal(token ? '0' : amount, feeInLamports));
+
+    if (totalSpent.isGreaterThan(availableBalance)) {
+        const error = token ? 'AMOUNT_NOT_ENOUGH_CURRENCY_FEE' : 'AMOUNT_IS_NOT_ENOUGH';
+        // errorMessage declared later
+        return { type: 'error', error, errorMessage: { id: error } } as const;
+    }
+
+    const payloadData = {
+        type: 'nonfinal',
+        totalSpent: token ? amount : totalSpent.toString(),
+        max,
+        fee: feeInLamports,
+        feePerByte: feeInLamports,
+        token,
+        bytes: 0,
+    } as const;
+
+    if (output.type === 'send-max' || output.type === 'external') {
+        return {
+            ...payloadData,
+            type: 'final',
+            // compatibility with BTC PrecomposedTransaction from @trezor/connect
+            transaction: {
+                inputs: [],
+                outputsPermutation: [0],
+                outputs: [
+                    {
+                        address: output.address,
+                        amount,
+                        script_type: 'PAYTOADDRESS',
+                    },
+                ],
+            },
+        };
+    }
+
+    return payloadData;
+};
+
+const fetchAccountOwnerAndTokenInfoForAddress = async (
+    address: string,
+    symbol: string,
+    mint: string,
+) => {
+    // Fetch data about recipient account owner if this is a token transfer
+    // We need this in order to validate the address and ensure transfers go through
+    let accountOwner: string | undefined;
+    let tokenInfo: TokenAccount[] | undefined;
+
+    const accountInfoResponse = await TrezorConnect.getAccountInfo({
+        coin: symbol,
+        descriptor: address,
+    });
+
+    if (accountInfoResponse.success) {
+        accountOwner = accountInfoResponse.payload?.misc?.owner;
+        tokenInfo = accountInfoResponse.payload?.tokens?.find(token => token.contract === mint)
+            ?.accounts;
+    }
+
+    return [accountOwner, tokenInfo] as const;
+};
+
+export const composeTransaction =
+    (formValues: FormState, formState: ComposeActionContext) =>
+    async (_dispatch: Dispatch, getState: GetState) => {
+        const { account, network, feeInfo } = formState;
+        const composeOutputs = getExternalComposeOutput(formValues, account, network);
+        if (!composeOutputs) return; // no valid Output
+
+        const { output, decimals, tokenInfo } = composeOutputs;
+
+        let fetchedFee: string | undefined;
+
+        const blockhash = getState().wallet.blockchain.sol.blockHash;
```

Comment:
Although I don't understand what does this blockhash parameter mean and why it differs every time I compose identical tx, I believe `DSOL` blockHash should be taken into account. But even if I replaced it, I wasn't able to send tx because of it.
![image](https://github.com/trezor/trezor-suite/assets/26326960/147dd21c-e92e-4281-9335-0156e64d80d0)

Reply:
Fixed in https://github.com/vacuumlabs/trezor-suite/commit/0c0891edaf73283e288b0caca2e23d914e0a5ad3. The problem was that the mainnet blockhash was always being added to the transaction - even on a devnet account.

The blockhash parameter is there to prevent replay attacks. A transaction will only be accepted if the blockhash in the transaction isn't older than X blocks/minutes - I think it's currently around 2 minutes. [Official docs here](https://docs.solana.com/developing/programming-model/transactions#recent-blockhash).